### ISSUE TITLE
undefined method `class_type?' for :included:Symbol

### ISSUE THREAD

**Issue by Wulfson** [bbatsov/rubocop#2471]
1 error occurred:
An error occurred while Style/MethodName cop was inspecting /home/jhancock/devops/dpn_libs/dpn/flow/baseflow.rb.
Errors are usually caused by RuboCop bugs.
Please, report your problems to RuboCop's issue tracker.
Mention the following information in the issue report:
0.35.1 (using Parser 2.2.3.0, running on ruby 1.9.3 x86_64-linux)

```
An error occurred while Style/MethodName cop was inspecting /home/jhancock/devops/dpn_libs/dpn/flow/baseflow.rb.
undefined method `class_type?' for :included:Symbol
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/mixin/configurable_naming.rb:37:in `block in class_emitter_method?'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/mixin/configurable_naming.rb:36:in `each'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/mixin/configurable_naming.rb:36:in `any?'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/mixin/configurable_naming.rb:36:in `class_emitter_method?'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/mixin/configurable_naming.rb:27:in `valid_name?'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/mixin/configurable_naming.rb:16:in `check_name'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/style/method_name.rb:19:in `on_defs'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:48:in `block (2 levels) in on_defs'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:87:in `with_cop_error_handling'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:47:in `block in on_defs'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:45:in `each'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:45:in `on_defs'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:259:in `process'
/var/lib/gems/1.9.1/gems/parser-2.2.3.0/lib/parser/ast/processor.rb:131:in `on_defs'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:52:in `on_defs'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:259:in `process'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:276:in `block in process_all'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `map'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `process_all'
/var/lib/gems/1.9.1/gems/parser-2.2.3.0/lib/parser/ast/processor.rb:9:in `process_regular_node'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:52:in `on_begin'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:259:in `process'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:276:in `block in process_all'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `map'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `process_all'
/var/lib/gems/1.9.1/gems/parser-2.2.3.0/lib/parser/ast/processor.rb:9:in `process_regular_node'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:52:in `on_module'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:259:in `process'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:276:in `block in process_all'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `map'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `process_all'
/var/lib/gems/1.9.1/gems/parser-2.2.3.0/lib/parser/ast/processor.rb:9:in `process_regular_node'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:52:in `on_module'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:259:in `process'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:276:in `block in process_all'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `map'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:275:in `process_all'
/var/lib/gems/1.9.1/gems/parser-2.2.3.0/lib/parser/ast/processor.rb:9:in `process_regular_node'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:52:in `on_module'
/var/lib/gems/1.9.1/gems/ast-2.1.0/lib/ast/processor/mixin.rb:259:in `process'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/commissioner.rb:53:in `investigate'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cop/team.rb:34:in `inspect_file'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:190:in `inspect_file'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:157:in `block in do_inspection_loop'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:151:in `loop'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:151:in `do_inspection_loop'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:93:in `process_file'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:58:in `block in inspect_files'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:56:in `each'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:56:in `inspect_files'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/runner.rb:34:in `run'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/lib/rubocop/cli.rb:26:in `run'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/bin/rubocop:13:in `block in <top (required)>'
/usr/lib/ruby/1.9.1/benchmark.rb:295:in `realtime'
/var/lib/gems/1.9.1/gems/rubocop-0.35.1/bin/rubocop:12:in `<top (required)>'
/usr/local/bin/rubocop:19:in `load'
/usr/local/bin/rubocop:19:in `<main>'
```

I assume it's choking on this method name:

```ruby
module DPN
	module Flow
		module BaseFlow
			include Autoload

			def self.included(base)
```

**Comment by alexdowad** [comment]
Can you post the entire source file which is causing the error? Because the excerpt which you posted does not cause the error, either on v0.35.1 or HEAD, using Ruby 1.9.3. If it's confidential, I can give you troubleshooting steps which you can perform on your side to pinpoint the cause of the error.

**Comment by Wulfson** [comment]
Yeah, I'll need to do the troubleshooting steps.

**Comment by alexdowad** [comment]
1. Clone this repo.
2. `cd` into the cloned project directory.
3. Copy the "problem" file into the project directory.
4. Check that the problem is repeatable using `ruby -I./lib bin/rubocop problem_file.rb`.
5. Edit `lib/rubocop/cop/mixin/configurable_naming.rb`. Just before line 36, add something like: `p node.parent`.
6. Try `ruby -I./lib bin/rubocop problem_file.rb` again, this time you should see some parenthesized expressions output at the console, showing the structure of the parse nodes which are being checked by `ConfigurableNaming#class_emitter_method?`.
7. The very last one will be the "problem" one. Cut-and-paste it into a Gist or something.

**Comment by Wulfson** [comment]
Ah, I was able to strip the file down, and the node output is actually quite small...the issue appears to be that there is a method defined within this method.

```
(defs
  (self) :included
  (args
    (arg :base))
  (defs
    (lvar :base) :Start
    (args
      (arg :aws_env)
      (restarg :args)) nil))
(defs
  (self) :included
  (args
    (arg :base))
  (defs
    (lvar :base) :Start
    (args
      (arg :aws_env)
      (restarg :args)) nil))
```

**Comment by Wulfson** [comment]
Here's a bare-bones example
```
module DPN
	module Flow
		module BaseFlow
			def self.included(base)
				def base.Start(aws_env, *args)

				end
			end
		end
	end
end
```

**Comment by alexdowad** [comment]
Do you want to code up a patch which fixes the problem, and submit it for inclusion in RuboCop? If so, I will guide you through the process. If you don't want to, then I will do it.

**Comment by Wulfson** [comment]
Feel free

**Comment by alexdowad** [comment]
This method is looking for things like:

```ruby
  module A
    class C; end
    def self.C; end
  end
```

The author expected the `def` nodes to be embedded in a class or module body, but not in the body of another method.

Do you see where it says `return false unless node.parent`? One simple fix would be to add a line after that which says something like: `return false if node.parent.defs_type?`

You'll also need to add a test case to catch regressions. There will probably be a spec file corresponding to `configurable_naming.rb`.

**Comment by Wulfson** [comment]
Sorry, I meant feel free to submit it...I don't know anything about the spec stuff.

Just to further constrain the actual issue, though, I stripped my bare-bones down more (updated above), and it looks like things do work properly if there's no error, or if there's no /\w\./ in front of the nested method's name. For example, my original example had `base.Start`, which threw the error, but if I remove the `base.` the error is not thrown. Similarly, if I changed the nested method name to `base.start` there was no error.

**Comment by alexdowad** [comment]
@Wulfson, do you have a class called `Start` somewhere in there as well?
